File: id/src/main/java/com/simprints/id/activities/RefusalActivity.java
package com.simprints.id.activities;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.KeyEvent;
import android.view.View;
import android.view.WindowManager;
import android.view.inputmethod.EditorInfo;
import android.widget.Button;
import android.widget.EditText;
import android.widget.RadioButton;
import android.widget.TextView;

import com.simprints.id.Application;
import com.simprints.id.R;
import com.simprints.id.data.DataManager;
import com.simprints.id.exceptions.unsafe.UninitializedDataManagerError;
import com.simprints.id.model.ALERT_TYPE;
import com.simprints.id.tools.AlertLauncher;
import com.simprints.id.tools.InternalConstants;
import com.simprints.id.libdata.models.enums.REFUSAL_FORM_REASON;
import com.simprints.libsimprints.Constants;
import com.simprints.libsimprints.RefusalForm;

public class RefusalActivity extends AppCompatActivity {

    private Button submit;
    private REFUSAL_FORM_REASON reason;
    private EditText otherText;
    private AlertLauncher alertLauncher;
    private DataManager dataManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_refusal);
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);

        Application app = ((Application) getApplication());
        dataManager = app.getDataManager();

        alertLauncher = new AlertLauncher(this);
        submit = findViewById(R.id.bt_submit_refusal_form);
        otherText = findViewById(R.id.et_other_refusal_text);

        otherText.clearFocus();

        submit.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                Intent resultData = new Intent();

                if (reason != null) {
                    RefusalForm refusalForm = new RefusalForm(reason.toString(), otherText.getText().toString());
                    try {
                        dataManager.saveRefusalForm(refusalForm);
                    } catch (UninitializedDataManagerError error) {
                        dataManager.logError(error);
                        alertLauncher.launch(ALERT_TYPE.UNEXPECTED_ERROR, 0);
                        return;
                    }
                    resultData.putExtra(Constants.SIMPRINTS_REFUSAL_FORM, refusalForm);
                }

                //We are really bailing out of Simprints ID so the result is 'cancelled'
                setResult(RESULT_CANCELED, resultData);
                finish();
            }
        });

        findViewById(R.id.bt_back_to_simprints).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                setResult(InternalConstants.RESULT_TRY_AGAIN);
                finish();
            }
        });

        otherText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                if (!otherText.getText().toString().isEmpty()) {
                    submit.setEnabled(true);
                }
            }
        });

        otherText.setOnEditorActionListener(new EditText.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                if (actionId == EditorInfo.IME_ACTION_DONE) {
                    submit.performClick();
                    return true;
                }
                return false;
            }
        });
    }

    public void onRefusalSelectionClicked(View view) {
        boolean checked = ((RadioButton) view).isChecked();

        if (checked) {
            otherText.setEnabled(true);

            if (view.getId() == R.id.rb_other && !otherText.getText().toString().isEmpty())
                submit.setEnabled(true);
            else
                submit.setEnabled(false);

            if (view.getId() != R.id.rb_other)
                submit.setEnabled(true);
        }

        switch (view.getId()) {
            case R.id.rb_scannerNotHere:
                if (checked)
                    reason = REFUSAL_FORM_REASON.SCANNER_NOT_HERE;
                break;
            case R.id.rb_scannerNotWorking:
                if (checked)
                    reason = REFUSAL_FORM_REASON.SCANNER_NOT_WORKING;
                break;
            case R.id.rb_unableToCapture:
                if (checked)
                    reason = REFUSAL_FORM_REASON.UNABLE_TO_CAPTURE_GOOD_SCAN;
                break;
            case R.id.rb_unableToGive:
                if (checked)
                    reason = REFUSAL_FORM_REASON.UNABLE_TO_GIVE_PRINTS;
                break;
            case R.id.rb_refused:
                if (checked)
                    reason = REFUSAL_FORM_REASON.REFUSED;
                break;
            case R.id.rb_other:
                if (checked)
                    reason = REFUSAL_FORM_REASON.OTHER;
                break;
        }
    }

}

File: id/src/main/java/com/simprints/id/activities/refusal/RefusalActivity.kt
package com.simprints.id.activities.refusal

import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.view.WindowManager
import androidx.appcompat.app.AppCompatActivity
import com.simprints.id.Application
import com.simprints.id.R
import com.simprints.id.domain.alert.Alert
import com.simprints.id.domain.refusal_form.RefusalFormAnswer
import com.simprints.id.domain.responses.RefusalFormResponse
import com.simprints.id.domain.responses.Response
import com.simprints.id.tools.extensions.launchAlert
import kotlinx.android.synthetic.main.activity_refusal.*
import org.jetbrains.anko.sdk27.coroutines.onLayoutChange

class RefusalActivity : AppCompatActivity(), RefusalContract.View {

    override lateinit var viewPresenter: RefusalContract.Presenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val component = (application as Application).component
        setContentView(R.layout.activity_refusal)
        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)

        viewPresenter = RefusalPresenter(this, component)

        setButtonClickListeners()
        setTextChangeListenerToRefusalText()
        setLayoutChangeListeners()
        setRadioGroupListener()
    }

    private fun setButtonClickListeners() {
        btSubmitRefusalForm.setOnClickListener { viewPresenter.handleSubmitButtonClick(getRefusalText()) }
        btScanFingerprints.setOnClickListener { viewPresenter.handleScanFingerprintsClick() }
    }

    private fun setTextChangeListenerToRefusalText() {
        refusalText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                viewPresenter.handleChangesInRefusalText(getRefusalText())
            }

            override fun afterTextChanged(s: Editable) {
            }
        })
    }

    //Changes in the layout occur when the keyboard shows up
    private fun setLayoutChangeListeners() {
        refusalScrollView.onLayoutChange { _, _, _, _,
                                           _, _, _, _, _ ->
            viewPresenter.handleLayoutChange()
        }
    }

    private fun setRadioGroupListener() {
        refusalRadioGroup.setOnCheckedChangeListener { _, optionIdentifier ->
            viewPresenter.handleRadioOptionClicked(optionIdentifier)
        }
    }

    override fun scrollToBottom() {
        refusalScrollView.post {
            refusalScrollView.fullScroll(View.FOCUS_DOWN)
        }
    }

    override fun enableSubmitButton() {
        btSubmitRefusalForm.isEnabled = true
    }

    override fun enableRefusalText() {
        refusalText.isEnabled = true
    }

    override fun doLaunchAlert(alert: Alert) {
        launchAlert(alert)
    }

    override fun setResultAndFinish(activityResult: Int, refusalAnswer: RefusalFormAnswer) {
        setResult(activityResult, getIntentForResultData(refusalAnswer))
        finish()
    }

    private fun getIntentForResultData(reason: RefusalFormAnswer) =
        Intent().putExtra(
            Response.BUNDLE_KEY,
            RefusalFormResponse(reason))

    private fun getRefusalText() = refusalText.text.toString()
}
